back-end/www/models/model_operations/location_operations.py

```diff
@@ -0,0 +1,278 @@
+from models.model import User
+from models.model_operations.user_operations import get_user_by_client_id
+from models.model_operations.answer_operations import get_answers_by_user
+
+
+DEBUG = False
+def dbprint(*values: object):
+    """
+    The printing function for debugging.
+    Enabled by the DEBUG flag.
+    """
+    if DEBUG:
+        print(values)
+
+
+def create_location(factory_id):
+    """
+    Create a location.
+
+    Parameters
+    ----------
+    factory_id : str
+        ID (uuid) provided by importing from the factory table.
+
+    Returns
+    -------
+    location : Location
+        The newly created location.
+    """
+    location = Location(factory_id=factory_id)
+
+    db.session.add(location)
+    db.session.commit()
+
+    return location
+
+
+def get_location_by_id(location_id):
+    """
+    Get a location by the location id.
+
+    Parameters
+    ----------
+    location_id : int
+        ID of the location.
+
+    Returns
+    -------
+    location : Location
+        The retrieved location object.
+    """
+    location = Location.query.filter_by(id=location_id).first()
+    return location
+
+
+def get_location_by_factory_id(factory_id):
+    """
+    Get a location by the factory id.
+
+    Parameters
+    ----------
+    factory_id : int
+        ID of the factory.
+
+    Returns
+    -------
+    location : Location
+        The retrieved location object.
+    """
+    location = Location.query.filter_by(factory_id=factory_id).first()
+    return location
+
+
+def set_location_done(location_id, is_done):
+    """
+    Set the current time to done_at to mark it is done.
+
+    Parameters
+    ----------
+    location_id : int
+        ID of the location.
+    is_done : bool
+        Set done or not done.
+
+    Returns
+    -------
+    location : Location
+        The retrieved location object.
+
+    Raises
+    ------
+    exception : Exception
+        When is_done is not bool
+    exception : Exception
+        When no location is found.
+    """
+    if not isinstance(is_done, bool):
+        raise Exception("is_done shall be bool")
+
+    location = Location.query.filter_by(id=location_id).first()
+
+    if location is None:
+        raise Exception("No location found in the database to update.")
+
+    if is_done:
+       location.done_at = datetime.datetime.now()
+    else:
+       location.done_at = None
+
+    db.session.commit()
+    return location
+
+
+def remove_location(location_id):
+    """
+    Remove a location.
+
+    Parameters
+    ----------
+    location_id : int
+        ID of the location.
+
+    Raises
+    ------
+    exception : Exception
+        When no location is found.
+    """
+    location = Location.query.filter_by(id=location_id).first()
+
+    if location is None:
+        raise Exception("No location found in the database to delete.")
+
+    db.session.delete(location)
+    db.session.commit()
+
+
+def get_locations(user_client_id, size, gold_standard_size):
```

Comment:
This is a complicated function. Make sure that you write test cases that cover different situations to check if the behavior of this function is correct.

Reply:
That's a good idea. Should I write one test case including all situations, or several test cases ?
@yalgorithm777  Thank you.

Reply:
You can put the test cases in the test_get_locations function in the location_tests.py file.

Reply:
I do include some exception testing, but can't think of more situations. Maybe for future improvement plan.